### DISCUSSION TITLE
Dump Controller State(DRO) and dump it into a file.

### DISCUSSION THREAD

**Discussion by pkr97** [winder/Universal-G-Code-Sender#2040] [Ideas]
We need a dump of controller state (workCoord machineCoord) with their respective timestamps. Could the community suggest a right way to tweak the existing code or if this functionality already exists in the current build (I couldn't find it in the GUI). 

I have installed Jdk-8.
I could compile Universal-G-Code-Sender-2.0.9.
I'm using apache-maven-3.8.6-bin & OpenJDK13U-jdk_x64_windows_hotspot_13.0.1_9 for compiling

Thanks for any help.

**Comment by breiler** [comment]
I think that the easiest way to do this is to query the pendant HTTP API.

Here is a search where you'll find information on how to get started:
https://github.com/winder/Universal-G-Code-Sender/search?q=pendant+api&type=discussions

**Reply by pkr97** [reply]
Hi Breiler,
Thanks and I could use pendent mode of communication.

Below is a code snippet I used to access the webpage and scraper to read Y-coordinate of the machine.

"@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
const puppeteer = require('puppeteer');

async function start() {
	const url = 'http://10.21.178.134:8080/';
	const browser = await puppeteer.launch({
		headless: false
	});
	const page = await browser.newPage();
	await page.goto(url);

	const dateObject = new Date();
	var yCord = 0.0
	
	var element = await page.waitForXPath("/html/body/app-root/div/div/div[1]/app-dro/div/div[2]/div[3]/div[2]")
	var yCord = await page.evaluate(element => element.textContent, element);
	console.log(yCord);
	
	while (1){
		var element = await page.waitForXPath("/html/body/app-root/div/div/div[1]/app-dro/div/div[2]/div[3]/div[2]")
		yCord = parseFloat(await page.evaluate(element => element.textContent, element));

		const timestamp = Date.now();
		console.log(timestamp, yCord);

		await page.waitForTimeout(1);
	}
	browser.close();
}
start()
"@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

However,
I see that the update rate of the pendant HTTP API is very slow and much less than 100Hz.

How do I move forward so I can record machine coordinates at a rate of 500Hz or 200Hz, if not 1000Hz?


Thanks for any help.

**Reply by breiler** [reply]
I think that it would be better by just query the API instead of scraping the web page: `curl -v -X GET http://localhost:8080/api/v1/status/getStatus`

The frequency for status updates is limited by the controller. UGS is polling the status by default every 200ms, there is a setting for making it more frequent. However, GRBL recommends a max status update frequency of 5-10Hz:
https://github.com/gnea/grbl/wiki/Grbl-v1.1-Interface#status-reporting

**Comment by chrisla23** [comment]
Eg:

curl -H "Content-Type: application/json" -X POST -d
"{\"commands\":\"G20G90G53G0Z0;G53G0X0Y0\"}"
http://192.168.7.85:8080/api/v1/machine/sendGcode

Also a handy one-liner:

FOR /F "delims=" %%A IN ('xidel.exe
http://192.168.7.196:8080/api/v1/status/getStatus -e
"rowCount:=($json).rowCount" -e
"completedRowCount:=($json).completedRowCount" -e
"remainingRowCount:=($json).remainingRowCount" -e
"sendDuration:=($json).sendDuration" -e
"sendRemainingDuration:=($json).sendRemainingDuration"  -e
"workCoordX:=round(($json).workCoord.x,2)" -e
"workCoordY:=round(($json).workCoord.y,2)" -e
"workCoordZ:=round(($json).workCoord.z,2)" -e "state:=($json).state" -e
"fileName:=($json).fileName" --output-format=cmd') DO %%A


-Chris
